File: Array/Medium/jump_game.py
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/jump-game/1/#
# Given an positive integer N and a list of N integers A[]. Each element in the array denotes
# the maximum length of jump you can cover. Find out if you can make it to the last index if
# you start at the first index of the list.
class Solution:
    def canReach(self, A, N):
        if N == 1:
            return 1
        
        reach = A[0]
        
        for i in range(1, N):
            if(reach < i):
                return 0
                
            if reach >= N: 
                return 1
                
            reach = max(reach , A[i] + i)
            
        return 1

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N=int(input())
        A=list(map(int,input().split()))
        
        ob = Solution()
        print(ob.canReach(A,N))
# } Driver Code Ends

File: Array/Medium/jump_game.cpp
// { Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
//User function Template for C++
// https://practice.geeksforgeeks.org/problems/jump-game/1/#
// Given an positive integer N and a list of N integers A[]. Each element in the array denotes
// the maximum length of jump you can cover. Find out if you can make it to the last index if
// you start at the first index of the list.
class Solution {
  public:
    int canReach(int A[], int N) {
        if(N==1) return true;
        int reach = A[0];
        
        for(int i = 1; i < N; i++)
        {
            if(reach < i) return false;
            if(reach >= N) return true;
            
            reach = max(reach, A[i] + i);
        }
        
        return true;
    }
};

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin>>N;
        
        int A[N];
        
        for(int i=0; i<N; i++)
            cin>>A[i];

        Solution ob;
        cout << ob.canReach(A,N) << endl;
    }
    return 0;
}  // } Driver Code Ends